### PR TITLE
Port aref

### PR DESCRIPTION
**Pull request by shaleh**
Implement Sub Char Table
Add Both a LispSubCharTableRef and a LispSubCharTableAsciiRef

Add more functionality to BoolVectors

Add tests for aref.

### CODE REVIEW

File: rust_src/remacs-sys/lib.rs [hunk 1]
```diff
@@ -765,14 +765,14 @@ extern "C" {
...
-#[allow(dead_code)]
-enum ChartabSize {
+pub enum ChartabSize {
     Bits0 = 6,
     Bits1 = 4,
     Bits2 = 5,
     Bits3 = 7,
 }
 
+/// Lisp_Char_Table
```
Comment (Wilfred) [inline rust_src/remacs-sys/lib.rs]: I'm not sure this comment adds anything :)

File: test/rust_src/src/data-tests.el [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                  :type 'args-out-of-range))
+  (should (eq (aref (kbd "<f1> SPC") 1)
+              32))
+  (let ((r (make-record 'foo 9 'Z)))
+    (should (eq (aref r 1)
+                'Z))
+    (should-error (aref r 10)
+                  :type 'args-out-of-range))
+  (should-error (aref 100 1)
+                :type 'wrong-type-argument))
```
Comment (Wilfred) [inline test/rust_src/src/data-tests.el]: Superb thoroughness :)

File: rust_src/src/hashtable.rs [hunk 3]
```diff
@@ -72,8 +73,10 @@ impl LispHashTableRef {
 
     #[inline]
     pub fn get_hash_value(self, idx: isize) -> LispObject {
-        let index = LispObject::from_natnum((2 * idx + 1) as EmacsInt);
-        unsafe { LispObject::from_raw(Faref(self.key_and_value, index.to_raw())) }
+        aref(
+            LispObject::from_raw(self.key_and_value),
+            (2 * idx + 1) as EmacsInt,
+        )
```
Comment (Wilfred) [inline rust_src/src/hashtable.rs]: It's nice to see how this is cleaning up our code and reducing `unsafe`.

File: rust_src/remacs-sys/lib.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/remacs-sys/lib.rs]: Following the existing style...

**Pull request merged**